### ISSUE TITLE
Can't interrupt select() or poll() calls from other threads (IDFGH-10532)

### ISSUE THREAD

**Issue by vlang-intona** [espressif/esp-idf#11777]
### Answers checklist.

- [X] I have read the documentation [ESP-IDF Programming Guide](https://docs.espressif.com/projects/esp-idf/en/latest/) and the issue is not addressed there.
- [X] I have updated my IDF branch (master or release) to the latest version and checked that the issue is present there.
- [X] I have searched the issue tracker for a similar issue and not found a similar issue.

### General issue report

select() and poll() wait for socket network events. But what if you need to respond to other events as well? Consider the following use case: an event triggered by another thread requires the select() thread to return from select() and to write a message to the socket. You would need a way to interrupt the select() or poll() call from another task. But there's nothing in esp-idf to do this. 

[FreeRTOS-Plus-TCP](https://www.freertos.org/FreeRTOS-Plus/FreeRTOS_Plus_TCP/index.html) has [FreeRTOS_SignalSocket()](https://www.freertos.org/FreeRTOS-Plus/FreeRTOS_Plus_TCP/API/FreeRTOS_SignalSocket.html), which is a nice possible solution.

Possible (non-)solutions I considered:

- use a timeout with select() (awful solution, because it increases both CPU use and latency)
- use a UNIX pipe or UNIX signals to unblock select() (esp-idf doesn't implement these)
- create local TCP or UDP dummy sockets and connect them to emulate the UNIX pipe trick (a bit awful, but a decent solution if you have existing code that absolutely requires such a mechanism)
- use different threads for reading and writing to a socket so you can do without select() (awful solution, requires 2 threads per socket and rewriting your code, probably more complicated code)
- use the netconn API (requires rewriting your code to use a LWIP specific API)

Possible solutions esp-idf could implement:

- add an API function similar to FreeRTOS_SignalSocket()
- add UNIX pipes or signals
- add another pseudo FD that exists only for signalling select() (look at eventfd() or signalfd() on Linux)

I think this would significantly improve the usefulness of the POSIX socket API on the ESP32.

One word about FreeRTOS_SignalSocket(): I didn't check whether it's sane. It looks somewhat similar to using UNIX signals to unblock select() calls - which is plagued by race condition issues. It could be that the FreeRTOS function might be affected by similar problems.

**Comment by igrr** [comment]
One solution for this problem already supported in ESP-IDF is eventfd (2): [docs](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/storage/vfs.html?highlight=eventfd#event-fds), [example](https://github.com/espressif/esp-idf/tree/master/examples/system/eventfd).

**Comment by vlang-intona** [comment]
Very nice. If this interoperates with sockets properly, this fully solves my problem.

**Comment by 0xjakob** [comment]
@vlang-intona Do you have any results already, i.e., does `eventfd()` solve your problem?

**Comment by vlang-intona** [comment]
I just ported my code from an awful contraption of multiple tasks to a single `poll()` loop with `eventfd()`. It seems to do what it's supposed to do. If anyone finds this post and wants to use `eventfd()`, don't forget to call `esp_vfs_eventfd_register()`. Also the include files are different from POSIX/Linux. I'm generally positively surprised by the ESP32 SDK.
